lib/grape/formatter/json_api_pagination.rb

```diff
@@ -0,0 +1,209 @@
+# frozen_string_literal: true
+
+module Grape
+  module Formatter
+    # This an extended and refactored version of the JSONAPI resources formatter
+    # that adds pagination links if there is a wrapper e.g. Kaminari that provides
+    # the appropriate pagination methods.
+    class JsonApiPagination
+      class << self
+        def call(resource, env)
+          new.serialize_resource(resource, env) || Grape::Formatter::Json.call(resource, env)
+        end
+      end
+
+      attr_reader :env, :resource, :endpoint
+
+      def serialize_resource(resource, env)
+        @env = env
+        @resource = resource
+        @endpoint = env['api.endpoint']
+
+        if resource_class.nil?
+          return blank_resource_response(jsonapi_options, resource)
+        end
+
+        build_main_json_output_data
+        add_pagination_links if resource_supports_pagination?
+
+        json_output.to_json
+      end
+
+      private
+
+      # Ensure sort order is maintained, serialize_to_hash can reorder objects if
+      # objects of the array are of different types (polymorphic cases)
+      def build_main_json_output_data
+        if sorted_primary_ids && (data = json_output['data']).present?
+          json_output['data'] = data.sort_by { |d| sorted_primary_ids.index(d['id']) }
+        end
+      end
+
+      def resource_supports_pagination?
+        resource.respond_to?(:current_page) && resource.respond_to?(:total_pages)
```

Comment:
Not sure how this formatter gets called, but would you want to raise if someone is trying to format with pagination when the resource doesn't support it?

Reply:
The idea is that it's optional. An active record scope without Kaminari should just get the JSONAPI stuff sent back with no pagination links.

Reply:
Got it. So it's active on all endpoints but only modifies the right ones here.

Reply:
Yeah